src/workerd/api/urlpattern.c++

```diff
@@ -4,309 +4,252 @@
-
-void URLPattern::visitForGc(jsg::GcVisitor& visitor) {
-  visitor.visit(protocolRegex, usernameRegex, passwordRegex, hostnameRegex, portRegex,
-      pathnameRegex, searchRegex, hashRegex);
+    if (value) {
+      result.groups.set(js, k, js.str(kj::ArrayPtr(value->c_str(), value->size())));
+    } else {
+      result.groups.set(js, k, js.undefined());
+    }
+  }
+  return result;
 }
 
-kj::StringPtr URLPattern::getProtocol() {
-  return inner.getProtocol().getPattern();
-}
-kj::StringPtr URLPattern::getUsername() {
-  return inner.getUsername().getPattern();
-}
-kj::StringPtr URLPattern::getPassword() {
-  return inner.getPassword().getPattern();
-}
-kj::StringPtr URLPattern::getHostname() {
-  return inner.getHostname().getPattern();
-}
-kj::StringPtr URLPattern::getPort() {
-  return inner.getPort().getPattern();
-}
-kj::StringPtr URLPattern::getPathname() {
-  return inner.getPathname().getPattern();
-}
-kj::StringPtr URLPattern::getSearch() {
-  return inner.getSearch().getPattern();
-}
-kj::StringPtr URLPattern::getHash() {
-  return inner.getHash().getPattern();
+bool URLPattern::getHasRegExpGroups() const {
+  return inner.has_regexp_groups();
 }
 
-URLPattern::URLPatternInit::operator jsg::UrlPattern::Init() {
-  return {
-    .protocol = this->protocol.map([](kj::String& str) { return kj::str(str); }),
-    .username = this->username.map([](kj::String& str) { return kj::str(str); }),
-    .password = this->password.map([](kj::String& str) { return kj::str(str); }),
-    .hostname = this->hostname.map([](kj::String& str) { return kj::str(str); }),
-    .port = this->port.map([](kj::String& str) { return kj::str(str); }),
-    .pathname = this->pathname.map([](kj::String& str) { return kj::str(str); }),
-    .search = this->search.map([](kj::String& str) { return kj::str(str); }),
-    .hash = this->hash.map([](kj::String& str) { return kj::str(str); }),
-    .baseUrl = this->baseURL.map([](kj::String& str) { return kj::str(str); }),
+URLPattern::URLPatternResult URLPattern::createURLPatternResult(
+    jsg::Lock& js, const ada::url_pattern_result& other) {
+  URLPatternResult result{
+#define V(_, name) .name = URLPattern::createURLPatternComponentResult(js, other.name),
+    URL_PATTERN_COMPONENTS(V)
+#undef V
   };
+
+  auto vecInputs = kj::heapArray<kj::OneOf<jsg::JsString, URLPatternInit>>(other.inputs.size());
+  size_t i = 0;
+  for (const auto& input: other.inputs) {
+    if (std::holds_alternative<std::string_view>(input)) {
+      auto raw = std::get<std::string_view>(input);
+      vecInputs[i] = js.str(kj::ArrayPtr(raw.data(), raw.size()));
+    } else {
+      KJ_DASSERT(std::holds_alternative<ada::url_pattern_init>(input));
+      auto obj = std::get<ada::url_pattern_init>(input);
+      vecInputs[i] = createURLPatternInit(obj);
+    }
+    i++;
+  }
+  result.inputs = mv(vecInputs);
+  return result;
 }
 
+#define V(uppercase, lowercase)                                                                    \
```

Comment:
Code like this is clearer if you give V a meaningful name like DEFINE_GETTER.

Reply:
Agreed. Updating now. Thanks for the review @erikcorry